netCDF4/_netCDF4.pyx

```diff
@@ -3754,27 +3757,64 @@ rename a `netCDF4.Variable` attribute named `oldname` to `newname`."""
+            # look for valid_min, valid_max.  No special
+            # treatment of byte data as described at
+            # http://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html).
+            if hasattr(self, 'valid_range') and len(self.valid_range) == 2:
+                validmin = numpy.array(self.valid_range[0], self.dtype)
+                validmax = numpy.array(self.valid_range[1], self.dtype)
+            else:
+                if hasattr(self, 'valid_min'):
+                    validmin = numpy.array(self.valid_min, self.dtype)
+                if hasattr(self, 'valid_max'):
+                    validmax = numpy.array(self.valid_max, self.dtype)
+            # http://www.unidata.ucar.edu/software/netcdf/docs/attribute_conventions.html).
+            # "If the data type is byte and _FillValue 
+            # is not explicitly defined,
+            # then the valid range should include all possible values.
+            # Otherwise, the valid range should exclude the _FillValue
+            # (whether defined explicitly or by default) as follows. 
+            # If the _FillValue is positive then it defines a valid maximum,
+            #  otherwise it defines a valid minimum."
+            byte_type = self.dtype.str[1:] in ['u1','i1']
+            if hasattr(self, '_FillValue'):
+                fval = numpy.array(self._FillValue, self.dtype)
+            else:
+                fval = numpy.array(default_fillvals[self.dtype.str[1:]],self.dtype)
+                if byte_type: fval = 0.
+            if validmin is None and fval < 0:
+                validmin  = fval
+            elif validmax is None and fval > 0:
+                validmax = fval
+            if validmin is not None:
+                mvalmask += data < validmin
+            if validmax is not None:
+                mvalmask += data > validmax
             if mval.shape == (): # mval a scalar.
-                hasmval = data==mval
+                mval = [mval] # make into iterable.
+            for m in mval:
                 # is scalar missing value a NaN?
                 try:
-                    mvalisnan = numpy.isnan(mval)
+                    mvalisnan = numpy.isnan(m)
                 except TypeError: # isnan fails on some dtypes (issue 206)
                     mvalisnan = False
-            else: # mval a vector.
-                hasmval = numpy.zeros(data.shape, numpy.bool)
-                for m in mval:
-                    m =  numpy.array(m)
-                    hasmval += data == m
-            if mval.shape == () and mvalisnan:
-                mask = numpy.isnan(data)
-            elif hasmval.any():
-                mask = hasmval
-            else:
-                mask = None
-            if mask is not None:
-                fill_value = mval
-                totalmask += mask
+                if mvalisnan: 
+                    mvalmask += numpy.isnan(data)
+                else:
+                    mvalmask += data==mval
```

Comment:
According to the conventions the missing value "... attribute is not treated in any special way by the library or conforming generic applications, but is often useful documentation and may be used by specific applications. The missing_value attribute can be a scalar or vector containing values indicating missing data. These values should all be outside the valid range so that generic applications will treat them as missing."

I for one would expect missing_values in the data to be masked whether inside the valid range or not - but that's not what the conventions appear to be saying. What's your take on this?